Implement API Diff workflow

S2Ler opened the pull request:
- API Diff job is only run on the correctly named PR branches.
- Depending on the release you are targeting, you need to append the
version in the first part of the branch name. See CONTRIBUTING.md for
more details.
There much limitation in CircleCI configs which make it difficult to
automate picking correct base API commit hash:
  - Can't have dynamic keys in the save_cache step.

High-level overview of the API Diff workflow:
1. Get API definition for the base API
2. Get API definition for the current API
3. Compare two APIs and report breaking changes

Other workflow improvements:
- Cache Carthage builds across workflow runs.

Authors: @Udumft @dersim-davaod @S2Ler

Udumft commented on CONTRIBUTING.md:
```diff
@@ -161,6 +161,12 @@ Pull requests are appreciated. If your PR includes any changes that would impact
 
+## Branch name conventions
+
+Each branch should start with the major version of the release the branch is targeting. For example:
```
Or here's another wild option for picking a base version:
What if we define multiple `type: approval` jobs, each one for a specific base version. If we need to compile an API diff for a specific version - we just "approve" the corresponding flow which configures the build steps with predefined base tags. Thus it is possible to maintain compatibility for old versions and well as append new releases easily.

S2Ler replied:
Thanks for good feedback. Below is my answers:

> Is it mandatory now to follow this pattern for each branch? 

- Good question. The answer is yes, and we can provide a step which will verify the branch name as an improvement. 

> I think that naming decision deserves a discussion

- The naming is up to discussion. Feel free to propose your variant with a motivation note. I picked to embed the version because it provides a way to see all v2 PRs in the first level (GUI git clients usually represent such branches as a path with folders). It is a more logical variant, in my opinion, but I don't have strong preferences. We can discuss the naming. 

> Also, I see that docs suggests trying v1

- [ ] v1 should be excluded from the docs for now because it is not a supported configuration. Thanks for the suggestion. 

> What if we define multiple type: approval jobs

- Having approval jobs pauses all the workflows until you explicitly accept it. That's why I rejected this variant. 

> Could we parse branch name to extract version number automatically

- I wanted to use automatic mapping from a target branch to base API commit but haven't found a way to do this due to CircleCI configuration limitations:
  - Cache keys can't have arbitrary data, and you can only use a limited set of env variables that are not suitable for our needs (see https://circleci.com/docs/2.0/configuration-reference/#savecache). 
  - There is no easy way to get the target branch of the PR. Even though this issue can be solved, the first limitation still blocks better implementation. 
If you have a suggestion how to implement this dynamic branching, let me know. 
  
> append new releases easily.

- Currently, we have only one supported version for the foreseeable future, so I don't think it makes sense to spend more time on optimizing adding new versions while there is no one on the horizon. 

> Thus it is possible to maintain compatibility for old versions

- We are following semantic versioning, and according to it, you shall introduce a major release when you have a breaking change. So, I don't think it makes sense to support such configuration in API diff workflow. Even if we need (very unlikely, in my opinion), we can add a new convention to the branch name or rethink this workflow.

azarovalex replied:
Does circle ci have any ability to access both _target_ and _base_ branches in the PR? We just need to make sure that there is no changes between this two.

S2Ler replied:
- There is no built-in functionality to access target branch.
- I haven't found a way to cache target branch API diffs if I get this target branch by other means. Cache keys are ver limiting.

S2Ler replied:
> Also, I see that docs suggests trying v1

Well, it turns out it is ok to keep the mention of v1 in CONTRIBUTING guide. It is just a convention. Whether CI will pick this convention isn't important here.

1ec5 replied:
> There is no built-in functionality to access target branch.

I’m surprised CircleCI doesn’t provide this functionality out of the box like other CI services do. I wonder if [this workaround](https://discuss.circleci.com/t/create-a-circle-target-branch-envar/10022/4) would be suitable.

1ec5 replied:
> Each branch should start with the major version of the release the branch is targeting.

This means if we ever decide to land a PR in a different major version, we can’t just rebase and retarget – we’d have to close the PR and open a new one. If it isn’t feasible to get the target branch, would it be feasible to get the milestone?

S2Ler replied:
> I’m surprised CircleCI doesn’t provide this functionality out of the box like other CI services do. I wonder if this workaround would be suitable.

>> Cache keys can't have arbitrary data, and you can only use a limited set of env variables that are not suitable for our needs (see https://circleci.com/docs/2.0/configuration-reference/#savecache).

S2Ler replied:
> This means if we ever decide to land a PR in a different major version, we can’t just rebase and retarget – we’d have to close the PR and open a new one.

Is it a problem to solve? What's bad about closing the PR? We already do this time to time. Just link the original PR if it has important conversation. 

>  If it isn’t feasible to get the target branch, would it be feasible to get the milestone?

The problem not that we can't get target branch, the problem is that we can't load cache for target branch, because we can't form the correct key. If we don't cache API results, it can be done.

1ec5 replied:
> Is it a problem to solve? What's bad about closing the PR? We already do this time to time. Just link the original PR if it has important conversation.

We actually move PRs from milestone to milestone quite often.

If the problem is specific to caching, what about a solution that requires us to do something explicit to cause the check to run? For example, I’ve seen CI integrations that require someone to post a comment on the PR with a special keyword that triggers a run of some expensive or noncritical check. I realize my suggestion would be scope-creep, but maybe we could live with not caching the results in the meantime. We did go without caching at all in this repository until #1632; the main downside was longer build times.

S2Ler replied:
I understand your concerns. I'm trying different approach which will be even better in my opinion.

S2Ler replied:
Fixed

S2Ler commented on .circleci/config.yml:
```diff
@@ -1,17 +1,141 @@
+          command: if [ ! -d /usr/local/Cellar/sourcekitten/<< parameters.sourcekitten-version >> ]; then brew update && brew install sourcekitten; fi
+      - save-sourcekitten-cache:
+          version: << parameters.sourcekitten-version >>      
+      - add_ssh_keys:
```
This key is only needed (and can only be used) to checkout `git@github.com:mapbox/swift-api-diff.git`. And the only reason to checkout this repository is to build it. And this does exactly this. If something will change, it is easy to refactor.

azarovalex commented on .circleci/config.yml:
```diff
@@ -1,17 +1,141 @@
+                key: nav-sdk-diff-cache-v5-<< parameters.key >>
 
+  build-api-diff-cli-tool:
+    description: "Builds CLI tool which produce diff reports"
```
Can we distribute `swift-api-diff` as a prebuild CLI so we don't have to build it each time?

azarovalex replied:
This will also allow to remove the steps that saves and restores the binary

S2Ler replied:
- Can be a subsequent improvement. Just one note: it is not built each time, it is built only once per released version.

S2Ler replied:
Feel free to post the suggestion in https://github.com/mapbox/swift-api-diff repo for future tracking or anywhere where it is more appropriate.

S2Ler commented on .circleci/config.yml:
```diff
@@ -295,8 +413,127 @@ jobs:
+      - *prepare-mapbox-file
+      - *prepare-netrc-file
+      - *update-carthage-version
+      - *add-github-to-known-hosts
```
fixed

S2Ler commented:
I found a way to do everything I wanted but without requiring branch convention. The idea is to have a file in the repository (`.base_api` at the moment) which will contain a hash of the commit which should be considered use for base API.

The caching is done by using `checksum` template component in cache key.

S2Ler commented:
@mapbox/navigation-ios I've addressed all the concerns. Let's land this PR right after we release RC to start tracking API breaking changes and find/fix edge cases if any.

Udumft commented on .base_api:
```diff
@@ -0,0 +1 @@
+e817932136f4f3894781bf32c5edf9b9863500d3
```
What would be the suggested flow when we do introduce the breaking changes?
For example, now that we've release rc.1, we can still do breaking changes in rc.2 so when should we update `.base_api`? If we do it in the same PR that does the breaking change - we won't see CI warning us and listing the breakages. But if we don't - we'll have to update `.base_api` in subsequent PR which is uncomfortable.

S2Ler replied:
Just update it in the same pr if breakage is approved

Udumft replied:
I would suggest then doing it in a separate commit, after CI does the run without such changes. It will allow to check if all the found breakages are intended and attach the report to the PR.

S2Ler replied:
That's reasonable. Good suggestion
